syntax = "proto3";
import "nanopb.proto";

enum FlightState {
    idle = 0;
    launchpad = 1;
    boost = 2;
    coast = 3;
    free_fall = 4;
    drogue_descent = 5;
    main_descent = 6;
    landed = 7;
}

enum FlightEvent {
    launch = 0;
    burnout = 1;
    apogee = 2;
    main_deploy = 3;
    drogue_deploy = 4;
    touchdown = 5;
    above_acs_threshold = 6;
}

message Acknowledge {
    bool success = 1;
}

message PressureReading {
    float pressure = 1;
}

message ImuReading {
    float ax = 1;
    float ay = 2;
    float az = 3;
    float gx = 4;
    float gy = 5;
    float gz = 6;
}

message StateEstimate {
    float x = 1;
    float y = 2;
    float z = 3;
    float vx = 4;
    float vy = 5;
    float vz = 6;

    float roll = 7;
    float pitch = 8;
    float yaw = 9;
}

message CANBus {
    bool can_started = 1;
    fixed32 loki_latest_rx_time = 2;
    fixed32 loki_latest_tx_time = 3;
    fixed32 sigurd_latest_rx_time = 4;
    fixed32 sigurd_latest_tx_time = 5;
}

message PyroStatus {
    bool pyro1_connected = 1;
    bool pyro2_connected = 2;
    bool pyro3_connected = 3;
}

message LokiToFjalar {
    int32 loki_state = 1;
    int32 loki_substate = 2;
    float loki_current_angle = 3;
    float loki_battery_voltage = 4;
}

message FjalarToLoki {
    FlightState flight_state = 1;
    FlightEvent flight_event = 2;
    bool enable_airbrakes = 3;
    float airbrakes_angle = 4;
}

message SigurdToFjalar{

}

message FjalarToSigurd{

}

message TelemetryPacket {
    float altitude = 1;
    float longitude = 2;
    float latitude = 3;
    bool pyro1_connected = 4;
    bool pyro2_connected = 5;
    bool pyro3_connected = 6;
    FlightState flight_state = 7;
    FlightEvent flight_event = 8;
    float az = 10;
    float velocity = 11;
    float battery = 12;
    int32 flash_address = 13;
    bool sudo = 14;
}

message GnssPosition {
    float longitude = 1;
    float latitude = 2;
    float altitude = 3;
}

message GnssStatus {
    int32 fix = 1;
    int32 num_satellites = 2;
    float hdop = 3;
    float vdop = 4;
}

message SetSudo {
    bool enabled = 1;
}

message ClearFlash {

}

message ReadFlash {
    int32 start_index = 1;
    int32 length = 2;
}

message FlashData {
    int32 start_index = 1;
    bytes data = 2 [(nanopb).max_size = 64];
}

message ReadyUp {

}

message EnterIdle {

}

message TriggerPyro {
    int32 pyro = 1;
}

message HilIn{
    float ax = 1;
    float ay = 2;
    float az = 3;

    float gx = 4;
    float gy = 5;
    float gz = 6;

    float p = 7;

    float lon = 8;
    float lat = 9;
    float alt = 10;

    uint32 time = 11;
}

message HilOut{
    float airbrake_percentage = 1;
    bool main_deployed = 2;
    bool drogue_deployed = 3;
}

message FjalarData {
    oneof data {
        Acknowledge acknowledge = 1;
        TelemetryPacket telemetry_packet = 2;
        ImuReading imu_reading = 3;
        PressureReading pressure_reading = 4;
        GnssPosition gnss_position = 5;
        GnssStatus gnss_status = 6;
        SetSudo set_sudo = 8;
        ClearFlash clear_flash = 9;
        ReadFlash read_flash = 10;
        FlashData flash_data = 11;
        ReadyUp ready_up = 12;
        EnterIdle enter_idle = 13;
        TriggerPyro trigger_pyro = 14;
        HilIn hil_in = 15;
        HilOut hil_out = 16;
        StateEstimate state_estimate = 17;
        FlightState flight_state = 18;
        FlightEvent flight_event = 19;
        CANBus can_bus = 20;
        PyroStatus pyro_status = 21;
        LokiToFjalar loki_to_fjalar = 22;
        FjalarToLoki fjalar_to_loki = 23;
        SigurdToFjalar sigurd_to_fjalar = 24;
        FjalarToSigurd fjalar_to_sigurd = 25;

    };
}

message FjalarMessage {
    fixed32 time = 1;
    int32 sequence_number = 2;
    FjalarData data = 3;
}

